File: 动态规划/背包问题/416.cpp
#include <vector>
using namespace std;
class Solution
{
public:
    bool canPartition(vector<int> &nums)
    {
        int Sum = 0;
        int MaxNum = 0;
        for (auto num : nums)
        {
            Sum += num;
            if (num > MaxNum)
                MaxNum = num;
        }
        if ((Sum % 2 != 0) || (MaxNum > Sum / 2) || (nums.size() < 2))
        {
            return false;
        }
        int target = Sum / 2;
        vector<vector<int>> dp(nums.size(), vector<int>(target + 1));
        dp[0][nums[0]] = 1;
        for (int i = 0; i < nums.size(); i++)
            dp[i][0] = 1;
        for (int i = 1; i < nums.size(); i++)
            for (int j = 0; j < target + 1; j++)
            {
                if(j>=nums[i])
                    dp[i][j] = dp[i-1][j]||dp[i-1][j-nums[i]];
                else
                    dp[i][j] = dp[i-1][j];
            }
        return dp[nums.size()-1][target];
    }
};

File: 动态规划/背包问题/416.py
from typing import List


class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        Sum = 0
        for num in nums:
            Sum += num
        if (Sum % 2 != 0) or len(nums)<2 or max(nums)>Sum//2:
            return False
        target = Sum//2
        dp = [[False for _ in range(target+1)] for _ in range(len(nums))]
        dp[0][nums[0]] = True
        for i in range(len(nums)):
            dp[i][0] = True
        for i in range(1,len(nums)):
            for j in range(target+1):
                if j>=nums[i]:
                    dp[i][j] = dp[i-1][j] or dp[i-1][j-nums[i]]
                else:
                    dp[i][j] = dp[i-1][j]
        return dp[len(nums)-1][target]


nums = [1,5,11,5]
obj = Solution()
print(obj.canPartition(nums))